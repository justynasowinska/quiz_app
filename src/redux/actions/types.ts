// questions
export const UPDATE_QUESTIONS = 'update_questions';

// progress
export const SET_NEXT_QUESTION = 'set_next_question';
export const RESET_PROGRESS = 'reset_progress';

// answers
export const ADD_USER_ANSWER = 'add_user_answer';
export const RESET_ANSWERS = 'reset_answers';