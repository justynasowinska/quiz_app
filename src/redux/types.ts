
export type DifficultlyType = 'hard' | 'medium' | 'easy';
export type BooleanCorrectAnswerType = 'True' | 'False';

export interface QuizQuestionType {
    category: string;
    type: string;
    difficulty: DifficultlyType;
    question: string;
    correct_answer: BooleanCorrectAnswerType; // maybe just string?
    incorrect_answers: Array<BooleanCorrectAnswerType>;
}

export type QuizQuestionsStateType = Array<QuizQuestionType>;

export interface QuizProgressType {
    currentQuestion: number;
}

export interface QuizAnswerType {
    questionId: number;
    answer: BooleanCorrectAnswerType;
}

export type QuizAnswersType = Array<QuizAnswerType>;