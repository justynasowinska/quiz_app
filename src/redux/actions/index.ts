export * from './answers';
export * from './progress';
export * from './questions';