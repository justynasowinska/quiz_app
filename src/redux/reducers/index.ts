import { combineReducers } from 'redux';

import quizDataReducer from './quizQuestionsReducer';
import quizProgressReducer from './quizProgressReducer';
import quizAnswersReducer from './quizAnswersReducer';

export const rootReducer = combineReducers({
    questions: quizDataReducer,
    progress: quizProgressReducer,
    answers: quizAnswersReducer
});

export default rootReducer;