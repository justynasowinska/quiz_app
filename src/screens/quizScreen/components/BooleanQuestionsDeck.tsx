import React, { useState } from 'react';
import { View, StyleSheet, PanResponder, Animated } from 'react-native';

import { QuizQuestionType, BooleanCorrectAnswerType } from '../../../redux/types';
import { QuestionCard } from './QuestionCard';
import { SWIPE_THRESHOLD, SWIPE_OUT_DURATION, ROTATE_CARD_DEG } from '../../../config/constants';
import { SCREEN_WIDTH } from '../../../utils/metrics';

interface PropsType {
    questions: ReadonlyArray<QuizQuestionType>;
    currentQuestionIndex: number;
    onUserAnswer: (questionIndex: number, answer: BooleanCorrectAnswerType) => void;
}

type SwipeDirectionType = 'right' | 'left';

export const BooleanQuestionsDeck = (props: PropsType) => {
    const [position] = useState(new Animated.ValueXY());

    const panResponder = PanResponder.create({
        onStartShouldSetPanResponder: () => true,
        onPanResponderMove: (_event, gesture) => {
            position.setValue({ x: gesture.dx, y: 0 });
        },
        onPanResponderRelease: (_event, gesture) => {
            if (gesture.dx > SWIPE_THRESHOLD) {
                forceSwipe('right');
            } else if (gesture.dx < -SWIPE_THRESHOLD) {
                forceSwipe('left');
            } else {
                resetPosition();
            }
        }
    });

    /**
     * Renders cards on deck. First card on the stack has panResponder callback attached.
     * To gain proper order we need to reverse array of cards, because cards are positioned absolute
     * and cover each other.
     */
    const renderCards = () => {
        const { questions, currentQuestionIndex } = props;

        return questions.map((question, index) => {
            if (index < currentQuestionIndex) {
                return null;
            }

            if (index === currentQuestionIndex) {
                return renderMoveableCard(question);
            }

            return renderNormalCard(question);
        }).reverse();
    };

    const renderMoveableCard = (question: QuizQuestionType) => (
        <Animated.View
            key={question.question}
            style={[getAnimatedCardStyle(), styles.cardStyle]}
            {...panResponder.panHandlers}
        >
            <QuestionCard
                question={question}
            />
        </Animated.View>
    );

    const renderNormalCard = (question: QuizQuestionType) => (
        <View
            key={question.question}
            style={styles.cardStyle}
        >
            <QuestionCard
                question={question}
            />
        </View>
    );

    const getAnimatedCardStyle = () => {
        const rotate = position.x.interpolate({
            inputRange: [-SCREEN_WIDTH * 1.5, 0, SCREEN_WIDTH * 1.5],
            outputRange: [`-${ROTATE_CARD_DEG}`, '0deg', ROTATE_CARD_DEG]
        });

        return {
            ...position.getLayout(),
            transform: [{
                rotate
            }]
        };
    };

    /**
     * Animates card out of the screen with given direction.
     * @param {SwipeDirectionType} direction  Direction in which card will be swiped off the screen.
     */
    const forceSwipe = (direction: SwipeDirectionType) => {
        const x = direction === 'right' ? SCREEN_WIDTH : -SCREEN_WIDTH;

        Animated.timing(position, {
            toValue: { x, y: 0 },
            duration: SWIPE_OUT_DURATION
        }).start(() => onSwipeComplete(direction));
    };

    const onSwipeComplete = (direction: SwipeDirectionType) => {
        const { onUserAnswer, currentQuestionIndex } = props;
        direction === 'right' ? onUserAnswer(currentQuestionIndex, 'True') : onUserAnswer(currentQuestionIndex, 'False');

        position.setValue({ x: 0, y: 0 });
    };

    const resetPosition = () => {
        Animated.spring(position, {
            toValue: { x: 0, y: 0 }
        }).start();
    };

    return (
        <View style={styles.container}>
            {renderCards()}
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        position: 'relative',
        width: '100%',
        aspectRatio: 1
    },
    cardStyle: {
        position: 'absolute'
    }
});