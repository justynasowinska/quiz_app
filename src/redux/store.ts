import { createStore, applyMiddleware } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/es/stateReconciler/autoMergeLevel2';
import storage from 'redux-persist/lib/storage';

import reducers from './reducers';

const persistConfig = {
    key: 'root',
    storage,
    stateReconciler: autoMergeLevel2
};

const persistedReducer = persistReducer(persistConfig, reducers);

const middlewares = [];

if (__DEV__) {
    const { logger } = require('redux-logger');

    middlewares.push(logger);
}

export const store = createStore(persistedReducer, applyMiddleware(...middlewares));
export const persistor = persistStore(store);