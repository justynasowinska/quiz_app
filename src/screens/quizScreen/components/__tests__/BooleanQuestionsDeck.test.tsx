import * as React from 'react';
import { shallow } from 'enzyme';

import { BooleanQuestionsDeck } from '../BooleanQuestionsDeck';
import { QuizQuestionType } from '../../../../redux/types';

describe('Quiz Screen BooleanQuestionsDeck Component', () => {
    it('matches to the snapshot', () => {
        const questions: ReadonlyArray<QuizQuestionType> = [{
            category: 'Entertainment: Video Games',
            type: 'boolean',
            difficulty: 'hard',
            question: 'Unturned originally started as a Roblox game.',
            correct_answer: 'True',
            incorrect_answers: ['False']
        }];

        const component = shallow(
            <BooleanQuestionsDeck
                questions={questions}
                currentQuestionIndex={0}
                onUserAnswer={jest.fn()}
            />
        );

        expect(component).toMatchSnapshot();
    });

    // TODO: Add more tests
});
