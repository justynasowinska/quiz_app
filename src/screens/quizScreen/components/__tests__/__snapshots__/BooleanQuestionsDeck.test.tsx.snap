// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Screen BooleanQuestionsDeck Component matches to the snapshot 1`] = `
<View
  style={
    Object {
      "aspectRatio": 1,
      "position": "relative",
      "width": "100%",
    }
  }
>
  <AnimatedComponent
    key="Unturned originally started as a Roblox game."
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Array [
        Object {
          "left": 0,
          "top": 0,
          "transform": Array [
            Object {
              "rotate": "0deg",
            },
          ],
        },
        Object {
          "position": "absolute",
        },
      ]
    }
  >
    <QuestionCard
      question={
        Object {
          "category": "Entertainment: Video Games",
          "correct_answer": "True",
          "difficulty": "hard",
          "incorrect_answers": Array [
            "False",
          ],
          "question": "Unturned originally started as a Roblox game.",
          "type": "boolean",
        }
      }
    />
  </AnimatedComponent>
</View>
`;
